File: gondola-core/rust/gondola-core/src/packets/telemetry_packet.rs
//! Wrapper for all telemetry data - original implementation in 
//! bfsw
// The following file is part of gaps-online-software and published 
// under the GPLv3 license

use crate::prelude::*;

/// A wrapper for packets from the telemetry stream
///
/// This is very compact and mostly used as an 
/// intermediary
#[derive(Debug, Clone, PartialEq)]
#[cfg_attr(feature = "pybindings", pyclass, pyo3(name="TelemetryPacket"))]
pub struct TelemetryPacket {
  pub header       : TelemetryPacketHeader,
  pub payload      : Vec<u8>,
  pub tof_paddles  : Arc<HashMap<u8,  TofPaddle>>, 
  pub trk_strips   : Arc<HashMap<u32, TrackerStrip>>,
}

#[cfg(feature="pybindings")]
#[pymethods]
impl TelemetryPacket {

  #[staticmethod]
  #[pyo3(name="get_gcutime_unpacked")]
  fn get_gcutime_unpacked_py(stream : Vec<u8>) -> PyResult<f64> {
    Ok(Self::get_gcutime_unpacked(&stream)?)
  }
  
  /// For a packet of any merged event type, retrieve the run id from 
  /// the TOF part 
  #[pyo3(name="get_runid")]
  fn get_runid_py(&self) -> PyResult<u16> {
    let runid_opt = self.get_runid();
    match runid_opt {
      Some(runid_res) => {
        match runid_res { 
          Ok(runid) => {
            return Ok(runid);
          }
          Err(err) => {
            return Err(PyValueError::new_err(err.to_string()));
          }
        }
      }
      None => {
        return Err(PyValueError::new_err("This packet does not seem to contain a (useful) runid!")); 
      }
    }
  }

  /// For a packet of type 80 (tracker standalooe) retrieve the GPS time from 
  /// the tracker header 
  #[pyo3(name="get_gpstime_tracker")]
  fn get_gpstime_tracker_py(&self) -> PyResult<u64> {
    let gpstime_opt = self.get_gpstime_tracker();
    match gpstime_opt {
      Some(gpstime_res) => {
        match gpstime_res { 
          Ok(gpstime) => {
            return Ok(gpstime);
          }
          Err(err) => {
            return Err(PyValueError::new_err(err.to_string()));
          }
        }
      }
      None => {
        return Err(PyValueError::new_err("This packet does not seem to contain a GPS time!")); 
      }
    }
  }
  
  /// In case this is any type of event packet which has a tof event, we can 
  /// also get the GPS time (as long as it is assigned) 
  #[pyo3(name="get_gpstime_tof")]
  fn get_gpstime_tof_py(&self) -> PyResult<u64> {
    let gpstime_opt = self.get_gpstime_tof();
    match gpstime_opt {
      Some(gpstime_res) => {
        match gpstime_res { 
          Ok(gpstime) => {
            return Ok(gpstime);
          }
          Err(err) => {
            return Err(PyValueError::new_err(err.to_string()));
          }
        }
      }
      None => {
        return Err(PyValueError::new_err("This packet does not seem to contain a GPS time!")); 
      }
    }
  }

  /// Get a zero copy view of the payload 
  /// Might be mostly useful for debugging purposes
  #[getter]
  fn payload<'py>(&self, py: Python<'py>) -> PyResult<Bound<'py, PyBytes>> {
    Ok(PyBytes::new(py, &self.payload))
  }

  #[getter]
  fn header(&self) -> TelemetryPacketHeader {
    // clone is fine here, since the packet header 
    // is pretty small
    self.header.clone()
  }

  #[getter]
  fn packet_type(&self) -> TelemetryPacketType {
    TelemetryPacketType::from(self.header.packet_type)
  }

  /// Check if this is either any of the different merged event 
  /// types 
  #[getter]
  #[pyo3(name="is_event_packet")]
  fn is_event_packet_py(&self) -> bool {
    self.is_event_packet()
  }

  /// Check if this packet is a complete run configuration 
  /// send by the TOF system
  #[getter] 
  #[pyo3(name="is_tof_toml_packet")]
  fn is_tof_toml_packet_py(&self) -> bool {
    if self.header.packet_type == TelemetryPacketType::AnyTofHK {
      // check the TOF packet type 
      // (self.payload contains a TOF packet) 
      let tof_packet_type = TofPacketType::from(parse_u8(&self.payload, &mut 2));
      return tof_packet_type == TofPacketType::LiftofSettings;
    }
    false
  }


  #[pyo3(name="to_bytestream")]
  fn to_bytestream_py(&self) -> Vec<u8> {
    self.to_bytestream()
  }

  #[staticmethod]
  #[pyo3(name="from_bytestream")]
  fn from_bytestream_py(stream : Vec<u8>, pos : usize) -> Result<Self, SerializationError> {
    let mut pos_ = pos;
    Self::from_bytestream(&stream, &mut pos_)
  }
}

impl TelemetryPacket {

  pub fn new() -> Self {
    Self {
      header      : TelemetryPacketHeader::new(),
      payload     : Vec::<u8>::new(),
      tof_paddles : Arc::new(HashMap::<u8, TofPaddle>::new()),
      trk_strips  : Arc::new(HashMap::<u32,TrackerStrip>::new()),
    }
  }
 
  pub fn is_event_packet(&self) -> bool {
    if self.header.packet_type == TelemetryPacketType::NoTofDataEvent
      || self.header.packet_type == TelemetryPacketType::NoGapsTriggerEvent 
      || self.header.packet_type == TelemetryPacketType::InterestingEvent 
      || self.header.packet_type == TelemetryPacketType::BoringEvent {
      true 
    } else {
      false
    }
  }

  /// Unpack the TelemetryPacket and return its content
  pub fn unpack<T>(&self) -> Result<T, SerializationError>
    where T: TelemetryPackable + Serialization {
    if !T::TEL_PACKET_TYPES_EVENT.contains(&self.header.packet_type) &&
      T::TEL_PACKET_TYPE != self.header.packet_type {
      error!("This bytestream is not for a {} packet!", self.header.packet_type);
      return Err(SerializationError::IncorrectPacketType);
    }
    let unpacked : T = T::from_bytestream(&self.payload, &mut 0)?;
    Ok(unpacked)
  }

  /// For a packet of any merged event type, retrieve the run id from 
  /// the TOF part 
  pub fn get_runid(&self) -> Option<Result<u16, SerializationError>> {
    if !self.is_event_packet() {
      return None;
    }
    if self.header.packet_type == TelemetryPacketType::NoTofDataEvent {
      return None;
    }
    // the run id is right after the timestamp (6 byte) 
    // and is 2 bytes long 
    if self.payload.len() < 37 {
      return Some(Err(SerializationError::StreamTooShort));
    }
    let mut pos   = 41usize;
    let runid = parse_u16(&self.payload, &mut pos);
    return Some(Ok(runid));
  }

  /// For a packet of type 80 (tracker standalooe) retrieve the GPS time from 
  /// the tracker header 
  pub fn get_gpstime_tracker(&self) -> Option<Result<u64, SerializationError>> {
    if self.header.packet_type != TelemetryPacketType::Tracker {
      return None;
    }
    // the tracker header with the gps time is the first 
    let mut pos = 10usize; // skip the first bytes in the TrackerHeader
    if self.payload.len() < 16 { 
      return Some(Err(SerializationError::StreamTooShort));
    }
    let lower = parse_u32(&self.payload, &mut pos);
    let upper = parse_u16(&self.payload, &mut pos);
    let ts    = make_systime(lower, upper);
    return Some(Ok(ts));
  }

  /// In case this is any type of event packet which has a tof event, we can 
  /// also get the GPS time (as long as it is assigned) 
  pub fn get_gpstime_tof(&self) -> Option<Result<u64, SerializationError>> {
    if !self.is_event_packet() {
      return None;
    }
    if self.header.packet_type == TelemetryPacketType::NoTofDataEvent {
      return None;
    }
    // then in the TelemetryEvent 
    // version (1byte)
    // flags0  (1byte)
    // -- only for version 1 supported 
    // + 8byte
    // event id (4byte)
    // tof dl   (1byte)
    // tof nby  (2byte)
    // -> 17 byte
    // ------ TofPacket 2 + 1 + 4 overhead (7byte) 
    // -> 24 byte 
    // TofEvent 2 + 1 + 2 + 1 + 4 + 1 
    // -> 35 byte 
    if self.payload.len() < 41 {
      return Some(Err(SerializationError::StreamTooShort));
    }
    //let mut pos = 42usize;
    let mut pos = 35usize;
    let ts32 = parse_u32(&self.payload, &mut pos);
    let ts16 = parse_u16(&self.payload, &mut pos);
    let ts   = 0x273000000000000 | (((ts16 as u64) << 32) | ts32 as u64);
    return Some(Ok(ts));
  }

  /// Get the gcutime from a packet without unpacking the full thing
  pub fn get_gcutime_unpacked(stream : &Vec<u8>) -> Result<f64, SerializationError> {
    // it starts with the serialized header and the packet byte, 
    // and then the timestamp is 32 bit. So we need to jump 3 bytes 
    // and then read 4 
    if stream.len() < 7 {
      error!("Can get gcutime from a bytestream shorter than 7 bytes!");
      return Err(SerializationError::StreamTooShort);
    }
    let mut pos = 3usize;
    let ts = parse_u32(stream, &mut pos);
    Ok(TelemetryPacketHeader::convert_telemetry_header_ts(ts))
  }

} 

impl Serialization for TelemetryPacket {

  /// No "classical" head byte marker
  const HEAD : u16 = 0;
  /// No "classical" tail byte marker
  const TAIL : u16 = 0;
  /// variable size
  const SIZE : usize = 0;

  fn from_bytestream(stream : &Vec<u8>, pos : &mut usize) -> Result<Self, SerializationError> {
    let mut tpacket: TelemetryPacket = TelemetryPacket::new();
    let header: TelemetryPacketHeader  = TelemetryPacketHeader::from_bytestream(stream, pos)?;
    tpacket.header = header;
    // to note here: The payload does not contain the payload for the header, it was just parsed
    tpacket.payload = stream[*pos..*pos + header.length as usize - TelemetryPacketHeader::SIZE].to_vec();
    Ok(tpacket)
  }

  fn to_bytestream(&self) -> Vec<u8> {
    let mut stream: Vec<u8> = Vec::<u8>::new();
    let mut s_head = self.header.to_bytestream();
    stream.append(&mut s_head);
    stream.extend_from_slice(self.payload.as_slice());
    stream
  }
}

impl Default for TelemetryPacket { 
  fn default() -> Self {
    Self::new()
  }
}

impl fmt::Display for TelemetryPacket {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    let mut repr: String = String::from("<TelemetryPacket:");
    repr += &(format!("\n  Header      : {}",self.header));
    repr += &(format!("\n  Payload len : {}>",self.payload.len()));
    write!(f, "{}", repr)
  }
}

impl Frameable for TelemetryPacket {
  const CRFRAMEOBJECT_TYPE : CRFrameObjectType = CRFrameObjectType::TelemetryPacket;
}


#[cfg(feature="pybindings")]
pythonize!(TelemetryPacket);

File: gondola-core/C++/src/telemetry_packet.cxx
/// This file is part of gaps-online-software and published 
/// under the GPLv3 license

#include "packets/telemetry_packet.hpp"
#include "io/parsers.h"

#include "spdlog/spdlog.h"
#include "spdlog/cfg/env.h"

namespace g   = gondola;

// make it more look like rust
using namespace result;

/// FIXME - this is obviously incomplete
u8 g::bfsw_ptype_to_u8(g::TelemetryPacketType pt) {
  switch (pt) {
    case g::TelemetryPacketType::Unknown:
      return 0;
    case g::TelemetryPacketType::Tracker:
      return 80;
    case g::TelemetryPacketType::BoringEvent:
      return 90;
    case g::TelemetryPacketType::RBWaveform:
      return 91;
    case g::TelemetryPacketType::InterestingEvent:
      return 190;
    case g::TelemetryPacketType::NoGapsTriggerEvent:
      return 191;
    case g::TelemetryPacketType::NoTofDataEvent:
      return 192;
    default:
      return 0;
  }
}

std::string g::bfsw_ptype_to_str(g::TelemetryPacketType pt) {
  switch (pt) {
      case g::TelemetryPacketType::Unknown:
      return "Unknown";
    case g::TelemetryPacketType::Tracker:
      return "Tracker";
    case g::TelemetryPacketType::BoringEvent:
      return "BoringEvent";
    case g::TelemetryPacketType::RBWaveform:
      return "RBWaveform";
    case g::TelemetryPacketType::InterestingEvent:
      return "InterestingEvent";
    case g::TelemetryPacketType::NoGapsTriggerEvent:
      return "NoGapsTriggerEvent";
    case g::TelemetryPacketType::NoTofDataEvent:
      return "NoTofDataEvent";
    case g::TelemetryPacketType::CoolingHK:
      return "CoolingHK";
    case g::TelemetryPacketType::CardHKP:
      return "CardHKP";
    case g::TelemetryPacketType::PDUHK:
      return "PDUHK";
    case g::TelemetryPacketType::AnyTofHK:
      return "AnyTofHK";
    case g::TelemetryPacketType::LabJackHK:
      return "LabJackHK";
    case g::TelemetryPacketType::TrackerDAQCntr:
      return "TrackerDAQCntr";
    default:
      return std::format("Unknown/NotImplemented ({})", (int)pt) ;
  }
}

//----------------------------------------

f64 g::TelemetryPacketHeader::get_gcutime() const {
  return timestamp * 0.064 + 1631030675.0;
};

//----------------------------------------

auto g::TelemetryPacketHeader::to_bytestream() const -> Vec<u8> {
  Vec<u8> bytes{0xeb, 0x90};
  bytes.push_back((u8)ptype);
  auto ts_byt = gondola::to_le_bytes(timestamp);
  auto co_byt = gondola::to_le_bytes(counter);
  auto le_byt = gondola::to_le_bytes(length);
  auto ch_byt = gondola::to_le_bytes(checksum);
  bytes.insert(bytes.end(),ts_byt.begin(), ts_byt.end());
  bytes.insert(bytes.end(),co_byt.begin(), co_byt.end());
  bytes.insert(bytes.end(),le_byt.begin(), le_byt.end());
  bytes.insert(bytes.end(),ch_byt.begin(), ch_byt.end());  
  return bytes;
}

auto  g::TelemetryPacketHeader::from_bytestream(Vec<u8> const &stream, usize &pos)
  -> Result<g::TelemetryPacketHeader, g::IOError> {
  g::TelemetryPacketHeader ph;
  if (stream.size() < pos + g::TelemetryPacketHeader::SIZE) {
    spdlog::error("The telemetry header is too short! ({} bytes when {} are expected!", stream.size(), g::TelemetryPacketHeader::SIZE);
    pos += g::TelemetryPacketHeader::SIZE;
    return Ok(ph);
  }
  if (g::parse_u16(stream, pos) != g::TelemetryPacketHeader::HEAD) {
    spdlog::error("The given position {} does not point to a valid header signature of {}", pos-2 ,g::TelemetryPacketHeader::HEAD);
    pos += g::TelemetryPacketHeader::SIZE - 2;
    return Ok(ph);
  }
  ph.sync      = g::TelemetryPacketHeader::HEAD;
  ph.ptype     = static_cast<g::TelemetryPacketType>(g::parse_u8 (stream, pos));
  ph.timestamp = g::parse_u32(stream, pos);
  ph.counter   = g::parse_u16(stream, pos);
  ph.length    = g::parse_u16(stream, pos);
  ph.checksum  = g::parse_u16(stream, pos);
  return Ok(ph);
}

auto g::TelemetryPacketHeader::to_string() const -> std::string {
  std::string repr = "<TelemetryPacketHeader:";
  repr += std::format("\n  Header      : {}" ,sync);
  repr += std::format("\n  Packet Type : {}" ,bfsw_ptype_to_str(ptype));
  repr += std::format("\n  Timestamp   : {}" ,timestamp);
  repr += std::format("\n  Counter     : {}" ,counter);
  repr += std::format("\n  Length      : {}" ,length);
  repr += std::format("\n  Checksum    : {}>",checksum);
  return repr;
}

//----------------------------------------

auto  g::TelemetryPacket::from_bytestream(Vec<u8> const &stream,
                                 usize &pos) -> g::TelemetryPacket {
  g::TelemetryPacket packet;
  auto header    = g::TelemetryPacketHeader::from_bytestream(stream, pos).unwrap();
  // FIXME
  packet.header  = header;
  auto payload   = g::slice(stream, pos, pos + header.length);
  packet.payload = std::move(payload);
  return packet;
}

auto g::TelemetryPacket::is_event_packet() const -> bool {
  return header.ptype == g::TelemetryPacketType::InterestingEvent   || 
         header.ptype == g::TelemetryPacketType::BoringEvent        || 
         header.ptype == g::TelemetryPacketType::NoGapsTriggerEvent || 
         header.ptype == g::TelemetryPacketType::NoTofDataEvent;      
}


auto g::TelemetryPacket::to_string() const -> std::string {
  std::string repr = "<TelemetryPacket:";
  repr += std::format("{}", header.to_string());
  repr += "\n --------";
  repr += std::format("\n  Payload len : {}>",payload.size());
  return repr;
}

